XRDDEV-1579 Autoscaler documentation (#57)
nordic-institute/X-Road-Security-Server-sidecar: doc/kubernetes_security_server_sidecar_user_guide.md

Edit the `values.yaml` file, set the `url` and `port` properties in the `prometheus` object with the following values:
``` yaml
[...]
# Url to access prometheus
prometheus:
 # Value is templated
 url: http://prometheus-operated.monitoring.svc.cluster.local
 port: 9090
 path: ""
[...]
```
The URL `prometheus-operated.monitoring.svc.cluster.local` is valid for AWS EKS clusters, for others scenarios the URL may be `prometheus-operated.monitoring.svc`

For detecting the network load, in this example, we are going to use the Prometheus metric `container_network_receive_packets_total` (We can use any other metric provided by Prometheus, like `container_network_receive_bytes_total` ... ). This metrics will inform us about the total number of packages received in our deployment. In principle, this metric may not be very useful to us, because it returns the total number of packets, but in our scenario, we will need to know the number packets in a certain time interval, for that reason, it is required to create a rule in the Prometheus Adapter configuration so that it returns the value of this metric based on the average of packets in a time interval.

> Edit the `values.yaml` file, creating a new custom rule:

Reviewer: ```suggestion
1. Edit the `values.yaml` file, creating a new custom rule:
```

``` yaml
[...]
rules:
  default: true
  custom:
  - seriesQuery: 'container_network_receive_packets_total{namespace="<namespace name>"}'
    resources:
      overrides:
        namespace:
          resource: namespace
        pod:
          resource: pod
        node:
          resource: node
    name:
      matches: "^(.*)_total"
      as: "${1}_per_minute"
    metricsQuery: 'sum(rate(<<.Series>>{<<.LabelMatchers>>}[1m])) by (<<.GroupBy>>)'
  # Mounts a configMap with pre-generated rules for use. Overrides the
  # default, custom, external and resource entries
  existing:
  external: []
[...]
```

>     * The `seriesQuery` property defines the metric and the namespace in wich this metric apply (the namespace of our Deployment). We can set `{namespace!=""}` if we want to make the metric available to all namespaces.

Reviewer: ```suggestion
    * The `seriesQuery` property defines the metric and the namespace in which this metric applies. Use the namespace of your Deployment if you want to make the metric available for your Deployment only or set `{namespace!=""}` if you want to make the metric available to all namespaces.
```

>     * The `resource` property defines wich type of Kubernetes resources can use this metric.

Reviewer: ```suggestion
    * The `resource` property defines which type of Kubernetes resources can use this metric.
```

    * The `name` property changes the metric name from `container_network_receive_packets_total` to `container_network_receive_packets_per_minute`.

>     * The `metricsQuery` property sum the average value of the metric in a time interval, in our example, 1 minute, but it could be any custom time interval.

Reviewer: ```suggestion
    * The `metricsQuery` property sums the average value of the metric in a time interval. For this example, the interval is set to 1 minute, but you can set any custom time interval.
```

More information about how to create rules can be found [here](https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/docs/config-walkthrough.md).

> Once the `values.yaml` file is configured, we can install the Prometheus Adapter by running (**reference data: 3.1**):

Reviewer: ```suggestion
2. Install the Prometheus Adapter by running (**reference data: 3.1**):
```

``` bash
helm install -f values.yaml  prometheus-adapter prometheus-community/prometheus-adapter --namespace <namespace name>
```

> After a couple of minutes, we can list the metrics, then verify that the metric `container_network_receive_packets_per_minute` we have just created is included in the list for pods resources by running:

Reviewer: ```suggestion
3. After a couple of minutes, list the metrics to verify that the metric `container_network_receive_packets_per_minute` created is included in the list for Pods resources:
```

``` bash
kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1 | jq
```

> If we need to modify the Prometheus Adapter rules once is deployed we can:

Reviewer: ```suggestion
If you need to modify the Prometheus Adapter rules once it's deployed, do the following:
```

>   * Edit the configuration file `values.yaml` and run a Helm upgrade (**reference data: 3.1**):

Reviewer: ```suggestion
  1. Edit the configuration file `values.yaml` and run a Helm upgrade (**reference data: 3.1**):
```

  ``` bash
  helm upgrade --install -f values.yaml  prometheus-adapter prometheus-community/prometheus --namespace <namespace name>
  ```

>   * Search for the Prometheus adapter ConfigMap and edit it (**reference data: 3.1**):

Reviewer: ```suggestion
  2. Search for the Prometheus adapter ConfigMap and edit it (**reference data: 3.1**):
```

  ``` bash
  kubectl get cm -n <namespace name>
  NAME                                                           DATA   AGE
  prometheus-adapter                                             1      4d5h
  prometheus-prometheus-kube-prometheus-prometheus-rulefiles-0   26     4d5h
  ```
  ``` bash
  kubectl edit cm -n <namespace name> prometheus-adapter
  ```

> It is possible uninstall the prometheus-adapter Helm chart by running (**reference data: 3.1**):

Reviewer: ```suggestion
You can uninstall the prometheus-adapter Helm chart by running (**reference data: 3.1**):
```

``` bash
helm uninstall prometheus-adapter -n <namespace name>
```

#### 9.4.3 Deploy HorizontalPodAutoescaler

Create the following manifets file (**reference data: 3.1, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6**) (The namespace should be the same than the Deployment namespace):
``` yaml
kind: HorizontalPodAutoscaler
apiVersion: autoscaling/v2beta1
metadata:
  name: <hpa name>
  namespace: <namespace name>
spec:
  scaleTargetRef:
    # point the HPA at the sample application
    # you created above
    apiVersion: apps/v1
    kind: Deployment
    name: <hpa deployment name>
  # autoscale between 1 and 10 replicas
  minReplicas: <hpa min replicas>
  maxReplicas: <hpa max replicas
  metrics:
  - type: Pods
    pods:
      metricName: <hpa metric name>
      # target 500 milli-requests per second,
      # which is 1 request every two seconds
      targetAverageValue: <target average value>
```

> Apply the HPA:

Reviewer: ```suggestion
2. Apply the HPA:
```

```
kubectl apply -f /path/to/hpa.yaml
```

>  If we describe the HPA we should see something similar to this (**reference data 3.1, 4.1**):

Reviewer: ```suggestion
3. If we describe the HPA we should see something similar to this (**reference data 3.1, 4.1**):
```

CreationTimestamp:                                         Tue, 02 Mar 2021 18:03:45 +0100
Reference:                                                 Deployment/secondary-sidecar-balancer-pod
Metrics:                                                   ( current / target )
  "container_network_receive_packets_per_minute" on pods:  5093m / 100
Min replicas:                                              1
Max replicas:                                              5
Deployment pods:                                           2 current / 2 desired
Conditions:
  Type            Status  Reason              Message
  ----            ------  ------              -------
  AbleToScale     True    ReadyForNewScale    recommended size matches current size
  ScalingActive   True    ValidMetricFound    the HPA was able to successfully calculate a replica count from pods metric container_network_receive_packets_per_minute
  ScalingLimited  False   DesiredWithinRange  the desired count is within the acceptable range
Events:           <none>

```

> The "Metrics" property give us information about the current value of the metric expressed in "milli" packets. In our example, it means that the Pods have recived 5093 "milli" packets in the last minute, and the target value is 100 (100000 milli packets), so the Pods will not scale up.

Reviewer: ```suggestion
The "Metrics" property gives information about the current value of the metric expressed in "milli" packets. In this example, it means that the Pods have received 5093 "milli" packets during the last minute. Given the target value of 100 (100000 milli packets), then the Pods will not scale up.
```

> Pods will scale up when the condition ( "current" > "target" * "Nº of replicas" ) is met.

Reviewer: ```suggestion
Pods will scale up only when the condition ( "current" > "target" * "Nº of replicas" ) is true.
```

> Now we can test to send request to our X-Road Security Server Sidecar cluster setup to see if the Pods are able to autoscale when the load increased.

Reviewer: ```suggestion
You can test sending requests to your X-Road Security Server Sidecar cluster setup to see if the Pods are able to autoscale when the network load increases.
```

### 9.5 Autoscale when Pods fails

> The HPA doesn't allow us to scale the Pods when one of them fails. But if we neet to always have a minumin number of Pods in our deployment and create a new one if one of them fails, we can combine the HPA with a [Liveness probe](#4552-liveness-probes).

Reviewer: ```suggestion
The HPA doesn't scale the Pods when one of them fails. But if you need to always have a minimum number of Pods in your Deployment so the HPA creates a new one if one of them fails, we can combine the HPA with a [Liveness probe](#4552-liveness-probes).
```

> Doing that we can set a minumin number of Pods in our HPA (**reference data: 4.2**) and configure a [Liveness probe](#4552-liveness-probes), that will attempt to recreate a new Pod if the healthcheck fails and the Pod is not able to receive messages.

Reviewer: ```suggestion
To do that, set a minimum number of Pods in your HPA (**reference data: 4.2**) and configure a [Liveness probe](#4552-liveness-probes) that will attempt to recreate a new Pod if the healthcheck fails and the Pod is not able to receive messages.
```

### 9.6 Installation with default metrics

> It is possible to skip the Prometheus and Prometheus Adapter steps and only use the default metrics provided by the [Metrics Server](https://github.com/kubernetes-sigs/metrics-server). With this metrics we can autoscale our Deployment base on CPU/Memory utilization.

Reviewer: ```suggestion
It is possible to skip the Prometheus and Prometheus Adapter steps and only use the default metrics provided by the [Metrics Server](https://github.com/kubernetes-sigs/metrics-server). With these metrics, you can autoscale your Deployment based on CPU/Memory utilization.
```

> For example, if we define a target of 50% percent of CPU utilization, the HPA will scale the number of replicas if the CPU utilization of any of the Pods in the deployment goes higher than 50%.

Reviewer: ```suggestion
For example, if we define a target value of 50% of CPU utilization, the HPA will scale up the number of replicas if the CPU utilization of any of the Pods in the Deployment goes over 50%.
```

> It is possible to create and HPA for CPU utilization by running (**reference data: 3.1, 4.2, 4.3, 4.5, 4.7**):

Reviewer: ```suggestion
Create an HPA for CPU utilization by running (**reference data: 3.1, 4.2, 4.3, 4.5, 4.7**):
```

``` bash
kubectl autoscale deployment <hpa deployment name> -n <namespace name> --cpu-percent=<target cpu percent> --min=<hpa min replicas> --max=<hpa max replicas>
```
The name of the HPA will be the same name as the Deployment.

> Another option for creating custom metrics is to create a manifest file, in this example we are going to show how to create an HPA based on Memory utilization (**reference data: 3.1, 4.1, 4.2, 4.3, 4.4, 4.5, 4.8**):

Reviewer: ```suggestion
Another option for creating custom metrics is to create a manifest file. The following example shows how to create an HPA based on Memory utilization (**reference data: 3.1, 4.1, 4.2, 4.3, 4.4, 4.5, 4.8**):
```

``` yaml

apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: <hpa name>
  namespace: <namespace name>
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: <hpa deployment name>
  minReplicas: <hpa min replicas>

>   maxReplicas: <hpa min replicas>

Reviewer: ```suggestion
  maxReplicas: <hpa max replicas>
```
